scripts/asr_language_modeling/ngram_lm/train_kenlm.py

```diff
@@ -54,95 +53,148 @@
-        model = nemo_asr.models.ASRModel.restore_from(args.nemo_model_file, map_location=torch.device('cpu'))
     else:
-        logging.warning(
-            "nemo_model_file does not end with .nemo, therefore trying to load a pretrained model with this name."
-        )
-        model = nemo_asr.models.ASRModel.from_pretrained(args.nemo_model_file, map_location=torch.device('cpu'))
-
-    encoding_level = kenlm_utils.SUPPORTED_MODELS.get(type(model).__name__, None)
-    if not encoding_level:
-        logging.warning(
-            f"Model type '{type(model).__name__}' may not be supported. Would try to train a char-level LM."
-        )
-        encoding_level = 'char'
-
-    """ DATASET SETUP """
-    logging.info(f"Encoding the train file '{args.train_file}' ...")
-    dataset = kenlm_utils.read_train_file(args.train_file, lowercase=args.do_lowercase)
-    encoded_train_file = f"{args.kenlm_model_file}.tmp.txt"
+        first_process_args = [
+            "python3",
+            os.path.join(file_path, "kenlm_utils.py"),
+            "--tokenizer_model_file",
+            args.tokenizer_model_file,
+            "--train_path",
+        ] + args.train_path
+        if args.rm_punctuation:
+            first_process_args.append("--rm_punctuation")
+        if args.do_lowercase:
+            first_process_args.append("--do_lowercase")
+        if args.separate_punctuation:
+            first_process_args.append("--separate_punctuation")
+
+    first_process = subprocess.Popen(first_process_args, stdout=subprocess.PIPE)
+
     if encoding_level == "subword":
-        kenlm_utils.tokenize_text(
-            dataset,
-            model.tokenizer,
-            path=encoded_train_file,
-            chunk_size=CHUNK_SIZE,
-            buffer_size=CHUNK_BUFFER_SIZE,
-            token_offset=DEFAULT_TOKEN_OFFSET,
-        )
-        # --discount_fallback is needed for training KenLM for BPE-based models
-        discount_arg = "--discount_fallback"
+        discount_arg = "--discount_fallback"  # --discount_fallback is needed for training KenLM for BPE-based models
     else:
-        with open(encoded_train_file, 'w', encoding='utf-8') as f:
-            for line in dataset:
-                f.write(f"{line}\n")
-
         discount_arg = ""
-
-    del model
-
     arpa_file = f"{args.kenlm_model_file}.tmp.arpa"
     """ LMPLZ ARGUMENT SETUP """
     kenlm_args = [
         os.path.join(args.kenlm_bin_path, 'lmplz'),
         "-o",
-        f"{args.ngram_length}",
-        "--text",
-        encoded_train_file,
+        str(args.ngram_length),
         "--arpa",
         arpa_file,
         discount_arg,
-    ]
+        "--prune",
+    ] + args.ngram_prun
 
-    ret = subprocess.run(kenlm_args, capture_output=False, text=True, stdout=sys.stdout, stderr=sys.stderr)
+    ret = subprocess.run(
+        kenlm_args, stdin=first_process.stdout, capture_output=False, text=True, stdout=sys.stdout, stderr=sys.stderr
```

Comment:
... ok this is pretty bad design. I'd rather not chain scripts like this. Can you make the original script a pure utils file, arg parse is fine but keep all args optional. Container can call arg parse, this script imports and calls things as needed via functions.

Reply:
fixed